File: bk/metadata/providers/goodreads.py
from datetime import datetime
import json
from bs4 import BeautifulSoup
import requests
from metadata.providers.base import MetadataProvider
from metadata.serializers import MetadataBookSerializer


class GoodreadsProvider(MetadataProvider):
    name = "Goodreads"
    URL: str = "https://www.goodreads.com/book/show/"

    def search(self, title: str) -> dict | None:
        r = requests.get(f"{self.URL}{title}")
        r.raise_for_status()

        soup = BeautifulSoup(r.text, "html.parser")
        script_tag = soup.find("script", id="__NEXT_DATA__")
        if not script_tag or not script_tag.string:
            return None

        data = json.loads(script_tag.string)
        apollo_state = data["props"]["pageProps"]["apolloState"]
        book_info = self._get_book_info(apollo_state)
        if not book_info:
            return None

        print(json.dumps(apollo_state, indent=4))

        book_dict = {
            "title": book_info.get("title", ""),
            "authors": self._extract_authors(apollo_state, book_info),
            "description": book_info.get("description", ""),
            "cover_image_url": book_info.get("imageUrl"),
            "asin": None,
            "isbn": None,
            "isbn13": None,
            "pages": None,
            "publication_time": None,
            "publisher": None,
            "language": None,
        }

        details = book_info.get("details", {})
        if details:
            ts = details.get("publicationTime")
            publication_time = datetime.fromtimestamp(ts / 1000) if ts else None
            lang = details.get("language") or {}

            book_dict.update(
                asin=details.get("asin"),
                isbn=details.get("isbn"),
                isbn13=details.get("isbn13"),
                pages=details.get("numPages"),
                publication_time=publication_time,
                publisher=details.get("publisher"),
                language=lang.get("name"),
            )

        serializer = MetadataBookSerializer(book_dict)
        print(json.dumps(serializer.data, indent=4))
        return serializer.data

    def _get_book_info(self, apollo_state: dict[str, dict]) -> dict | None:
        for key, val in apollo_state.items():
            if key.lower().startswith("book:"):
                return val
        return None

    def _get_primary_author_ref(self, book_info: dict) -> str | None:
        edge = book_info.get("primaryContributorEdge")
        if not edge:
            return None
        node = edge.get("node")
        if not node:
            return None
        return node.get("__ref")

    def _resolve_author_key(self, ref: str | None, apollo_state: dict) -> str | None:
        if not ref:
            return None
        if ref in apollo_state:
            return ref
        return None

    def _extract_authors(
        self, apollo_state: dict[str, dict], book_info: dict
    ) -> list[dict]:
        authors = []

        primary_ref = self._get_primary_author_ref(book_info)
        primary_key = self._resolve_author_key(primary_ref, apollo_state)

        if primary_key:
            author_details = apollo_state.get(primary_key)
            if author_details:
                authors.append(
                    {
                        "name": author_details.get("name", ""),
                        "bio": author_details.get("description") or "",
                    }
                )

        for edge in book_info.get("secondaryContributorEdges", []):
            if edge.get("role", "").lower() != "author":
                continue

            node_ref = edge.get("node", {}).get("__ref")
            key = self._resolve_author_key(node_ref, apollo_state)
            if not key:
                continue

            author_details = apollo_state.get(key)
            if not author_details or not author_details.get("name"):
                continue

            authors.append(
                {
                    "name": author_details.get("name", ""),
                    "bio": author_details.get("description") or "",
                    "is_primary": False,
                }
            )

        return authors

File: crates/shiori_metadata/src/goodreads.rs
use crate::{errors::MetadataError, provider::MetadataProvider};
use chrono::DateTime;
use serde_json::Value;
use shiori_api_types::EncodableMetadataSearch;

pub struct GoodreadsProvider;

// TODO: Def refactor this, i dont like how it looks rn lol
//
// Would prob be a good idea to make my own types for this.
impl MetadataProvider for GoodreadsProvider {
    const URL: &str = "https://www.goodreads.com/book/show/";

    async fn search(id: &str) -> Result<EncodableMetadataSearch, MetadataError> {
        let body = reqwest::get(format!("{}{}", Self::URL, id))
            .await?
            .error_for_status()?
            .text()
            .await?;

        let document = scraper::Html::parse_document(&body);
        let selector = scraper::Selector::parse("script#__NEXT_DATA__")
            .map_err(|_| MetadataError::HtmlParse)?;

        let script = document
            .select(&selector)
            .next()
            .ok_or_else(|| MetadataError::MissingScriptTag)?;

        let next_data: Value = serde_json::from_str(&script.inner_html())?;

        let apollo_state = &next_data["props"]["pageProps"]["apolloState"];

        let book_info =
            Self::book_info(apollo_state).ok_or_else(|| MetadataError::MissingBookInfo)?;

        let mut metadata = EncodableMetadataSearch {
            title: book_info
                .get("title")
                .and_then(Value::as_str)
                .map(String::from)
                .unwrap_or_default(),
            authors: Self::extract_author_names(apollo_state, book_info),
            cover_url: book_info
                .get("imageUrl")
                .and_then(Value::as_str)
                .map(String::from),
            description: book_info
                .get("description")
                .and_then(Value::as_str)
                .map(String::from),
            genres: Self::extract_genres(book_info),
            ..Default::default()
        };

        if let Some(details) = book_info.get("details") {
            metadata.publisher = details
                .get("publisher")
                .and_then(Value::as_str)
                .map(String::from);

            metadata.isbn = details
                .get("isbn")
                .and_then(Value::as_str)
                .map(String::from);

            metadata.language = details
                .get("language")
                .and_then(|l| l.get("name"))
                .and_then(Value::as_str)
                .map(String::from);

            metadata.published_at = details
                .get("publicationTime")
                .and_then(|time| time.as_i64())
                .and_then(DateTime::from_timestamp_millis)
                .map(|datetime| datetime.date_naive())
        }

        Ok(metadata)
    }
}

// TODO: maybe extract people like translators, illustrator, etc.
impl GoodreadsProvider {
    /// Extract the first key starting with "book:" and return its value
    fn book_info(apollo_state: &Value) -> Option<&Value> {
        apollo_state
            .as_object()?
            .iter()
            .find(|(k, _)| k.to_lowercase().starts_with("book:"))
            .map(|(_, v)| v)
    }

    /// Return a list of genres
    fn extract_genres(book_info: &Value) -> Vec<String> {
        book_info
            .get("bookGenres")
            .and_then(|g| g.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|g| {
                        g.get("genre")
                            .and_then(|g| g.get("name"))
                            .and_then(Value::as_str)
                            .map(|s| s.to_string())
                    })
                    .collect()
            })
            .unwrap_or_default()
    }

    /// Given a reference key, return the author's name if it exists
    fn get_author_name<'a>(apollo_state: &'a Value, r: Option<&'a str>) -> Option<&'a str> {
        let key = r?;
        apollo_state.as_object()?.get(key)?.get("name")?.as_str()
    }

    /// Return a list of all author names
    fn extract_author_names(apollo_state: &Value, book_info: &Value) -> Vec<String> {
        let mut authors = Vec::new();

        // Primary author
        let primary_ref = book_info
            .get("primaryContributorEdge")
            .and_then(|e| e.get("node"))
            .and_then(|n| n.get("__ref"))
            .and_then(Value::as_str);

        if let Some(name) = Self::get_author_name(apollo_state, primary_ref) {
            authors.push(name);
        }

        // Secondary authors
        if let Some(edges) = book_info
            .get("secondaryContributorEdges")
            .and_then(Value::as_array)
        {
            for edge in edges {
                let is_author = edge
                    .get("role")
                    .and_then(Value::as_str)
                    .map(|r| r.eq_ignore_ascii_case("author"))
                    .unwrap_or(false);

                if !is_author {
                    continue;
                }

                let node_ref = edge
                    .get("node")
                    .and_then(|n| n.get("__ref"))
                    .and_then(Value::as_str);
                if let Some(name) = Self::get_author_name(apollo_state, node_ref) {
                    authors.push(name);
                }
            }
        }

        authors.iter().map(|s| s.to_string()).collect()
    }
}
